server/text_generation_server/utils/weights.py

```diff
@@ -1,13 +1,66 @@
-import os
+from abc import ABC, abstractmethod
 from pathlib import Path
 from typing import Dict, List, Optional, Union
-from safetensors import safe_open, SafetensorError
+from safetensors import safe_open
 import torch
-from loguru import logger
-from huggingface_hub import hf_hub_download
-import json
-from text_generation_server.layers.gptq import GPTQParams
-from text_generation_server.utils.log import log_once
+
+
+class WeightsLoader(ABC):
+    """
+    Instances of this type implement higher-level weight loading.
+
+    At a low-level, every weight is stored in the Safetensors format.
+    The interpretation of weights may be different however, for instance
+    could be packed, quantized weights. Loaders are responsible for
+    interpreting the raw tensors, sharding tensors in a manner compatible
+    with the format, etc.
+    """
+
+    @abstractmethod
+    def get_weights_col_packed(
```

Comment:
I think all from this interface could use a bit of doc.


imho there's only 3 methods.


col_packed -  (Q,  K, V are concatenated in the model, therefore with TP we need to know that it's the case to slice into that tensor correctly).

col_multi - Q, K, V are split, we need to know each name to slice properly and fuse everything into a single tensor


row - (Row is currently never packed nor multi therefore a simple sharding algorithm, suffice.

Am I missing something ?

Reply:
Added docs.